pageserver/src/virtual_file/owned_buffers_io/write/flush.rs

```diff
@@ -0,0 +1,200 @@
+}
+
+impl<Buf, W> FlushBackgroundTask<Buf, W>
+where
+    Buf: IoBufAligned + Send + Sync,
+    W: OwnedAsyncWriter + Sync + 'static,
+{
+    /// Creates a new background flush task.
+    fn new(
+        channel: Duplex<FullSlice<Buf>, (FullSlice<Buf>, u64)>,
+        file: Arc<W>,
+        ctx: RequestContext,
+    ) -> Self {
+        FlushBackgroundTask {
+            channel,
+            writer: file,
+            ctx,
+        }
+    }
+
+    /// Runs the background flush task.
+    async fn run(mut self, slice: FullSlice<Buf>) -> std::io::Result<Arc<W>> {
+        // Sends the extra buffer back to the handle.
+        self.channel.send(slice).await.map_err(|_| {
+            std::io::Error::new(std::io::ErrorKind::BrokenPipe, "flush handle closed early")
+        })?;
+
+        //  Exit condition: channel is closed and there is no remaining buffer to be flushed
+        while let Some((slice, offset)) = self.channel.recv().await {
+            // Write slice to disk at `offset`.
+            let slice = self.writer.write_all_at(slice, offset, &self.ctx).await?;
+
+            // Sends the buffer back to the handle for reuse. The handle is in charged of cleaning the buffer.
+            if self.channel.send(slice).await.is_err() {
+                // Although channel is closed. Still need to finish flushing the remaining buffers.
+                continue;
+            }
+        }
+
+        Ok(self.writer)
+    }
+}
+
+impl<Buf, W> FlushHandle<Buf, W>
+where
+    Buf: IoBufAligned + Send + Sync + Clone,
+    W: OwnedAsyncWriter + Send + Sync + 'static + std::fmt::Debug,
+{
+    /// Spawns a new background flush task and obtains a handle.
+    pub fn spawn_new<B>(file: Arc<W>, buf: B, ctx: RequestContext) -> Self
+    where
+        B: Buffer<IoBuf = Buf> + Send + 'static,
+    {
+        let (front, back) = duplex_channel(2);
+
+        let bg = FlushBackgroundTask::new(back, file, ctx);
+        let join_handle = tokio::spawn(async move { bg.run(buf.flush()).await });
+
+        FlushHandle {
+            inner: Some(FlushHandleInner {
+                channel: front,
+                join_handle,
+            }),
+            maybe_flushed: None,
+        }
+    }
+
+    /// Submits a buffer to be flushed in the background task.
+    /// Returns a buffer that completed flushing for re-use, length reset to 0, capacity unchanged.
+    /// If `save_buf_for_read` is true, then we save the buffer in `Self::maybe_flushed`, otherwise
+    /// clear `maybe_flushed`.
+    pub async fn flush<B>(
+        &mut self,
+        buf: B,
+        offset: u64,
+        save_buf_for_read: bool,
+    ) -> std::io::Result<B>
+    where
+        B: Buffer<IoBuf = Buf> + Send + 'static,
+    {
+        let freezed = buf.flush();
+
+        // Saves a buffer for read while flushing. This also removes reference to the old buffer.
+        self.maybe_flushed = if save_buf_for_read {
+            Some(freezed.as_raw_slice().get_ref().clone())
```

Comment:
robustness: if ever there's a bug or future change where the slice bounds aren't `0..$len_of_the_Bytes` then we risk giving acess to stale data here.

Can we do better? E.g., a specialized impl of `FullSlice<Bytes>` that [does `Bytes::slice()` under the hood](https://docs.rs/bytes/latest/bytes/struct.Bytes.html#method.slice)?

Reply:
d4ebd5ccd31e61a1c2aca637f058fac74cf5cb40